france_connect/clients.py

```diff
@@ -0,0 +1,222 @@
+        fc_authorization_url: str = "/api/v2/authorize",
+        fc_token_url: str = "/api/v2/token",
+        fc_jwks_url: str = "/api/v2/jwks",
+        fc_userinfo_url: str = "/api/v2/userinfo",
+        fc_logout_url: str = "/api/v2/session/end",
+        timeout: int = 10,
+        verify_ssl: bool = True,
+        allow_redirects: bool = True,
+    ):
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.scopes = scopes
+        self.post_login_callback_url = post_login_callback_url
+        self.post_logout_callback_url = post_logout_callback_url
+        self.post_contact_callback_url = post_contact_callback_url
+        self.fc_base_url = fc_base_url
+        self.fc_authorization_url = fc_authorization_url
+        self.fc_token_url = fc_token_url
+        self.fc_jwks_url = fc_jwks_url
+        self.fc_userinfo_url = fc_userinfo_url
+        self.fc_logout_url = fc_logout_url
+        self.timeout = timeout
+        self.verify_ssl = verify_ssl
+        self.allow_redirects = allow_redirects
+
+    @staticmethod
+    def generate_nonce() -> str:
+        """Generate a random nonce in hexadecimal string (64 characters)."""
+        return secrets.token_hex(64)
+
+    @staticmethod
+    def generate_state() -> str:
+        """Generate a random state in hexadecimal string (64 characters)."""
+        return secrets.token_hex(64)
+
+    def get_authentication_url(
+        self, acr_values: List[ACRValues] = None, nonce: str = None, state: str = None
+    ) -> Tuple[str, str, str]:
+        """Return the authentication URL, nonce, and state.
+
+        Parameters
+        ----------
+        acr_values: List[ACRValues], optional
+            The list of ACR values used for the authentication process, defaults
+            to `[ACRValues.EIDAS1]`.
+        nonce: str, optional
+            The `nonce` used for the authentication process, generated using
+            `generate_nonce()` if not provided.
+        state: str, optional
+            The `state` used for the authentication process, generated using
+            `generate_state()` if not provided.
+
+        Returns
+        -------
+        Tuple[str, str, str]
+            The authentication URL, `nonce` and `state`.
+        """
+        acr_values = acr_values or [ACRValues.EIDAS1]
+        nonce = nonce or self.generate_nonce()
+        state = state or self.generate_state()
+        params = urllib.parse.urlencode(
+            {
+                "client_id": self.client_id,
+                "response_type": "code",
+                "scope": " ".join(self.scopes),
+                "acr_values": " ".join(acr_values),
+                "nonce": nonce,
+                "state": state,
+                "redirect_uri": self.post_login_callback_url,
+            }
+        )
+        url = urllib.parse.urljoin(self.fc_base_url, self.fc_authorization_url)
+        return f"{url}?{params}", nonce, state
+
```

Comment:
We should have a `login_callback_url=None` which default to `self.login_callback_url`.